core/bin/zksync_tee_prover/src/tee_prover.rs

```diff
@@ -0,0 +1,142 @@
+        let tee_prover_task = TeeProver {
+            signing_key: self.signing_key,
+            attestation_quote_bytes: self.attestation_quote_bytes,
+            tee_type: self.tee_type,
+            api_client: ApiClient::new(self.api_url, Client::new()),
+        };
+        context.add_task(Box::new(tee_prover_task));
+        Ok(())
+    }
+}
+
+struct TeeProver {
+    signing_key: SecretKey,
+    attestation_quote_bytes: Vec<u8>,
+    tee_type: TeeType,
+    api_client: ApiClient,
+}
+
+impl TeeProver {
+    fn verify(&self, tvi: TeeVerifierInput) -> anyhow::Result<(Signature, L1BatchNumber, H256)> {
+        match tvi.verify() {
+            Err(e) => {
+                let err_msg = format!("L1 batch verification failed: {e}");
+                tracing::warn!(err_msg);
+                Err(anyhow::anyhow!(err_msg))
+            }
+            Ok(verification_result) => {
+                let root_hash_bytes = verification_result.0.as_bytes();
+                let batch_number = verification_result.1;
+                let msg_to_sign = Message::from_slice(root_hash_bytes)?;
+                let signature = self.signing_key.sign_ecdsa(msg_to_sign);
+                Ok((signature, batch_number, verification_result.0))
+            }
+        }
+    }
+}
+
+#[async_trait::async_trait]
+impl Task for TeeProver {
+    fn id(&self) -> TaskId {
+        "tee_prover".into()
+    }
+
+    async fn run(self: Box<Self>, mut stop_receiver: StopReceiver) -> anyhow::Result<()> {
+        tracing::info!("Starting the task {}", self.id());
+
+        self.api_client
+            .register_attestation(self.attestation_quote_bytes.clone(), &self.signing_key)
+            .await?;
+
+        const POLLING_INTERVAL_MS: u64 = 1000;
+        const MAX_BACKOFF_MS: u64 = 60_000;
+        const BACKOFF_MULTIPLIER: u64 = 2;
+
+        let mut backoff: u64 = POLLING_INTERVAL_MS;
+
+        loop {
+            if *stop_receiver.0.borrow() {
+                tracing::warn!("Stop signal received, shutting down TEE Prover component");
+                return Ok(());
+            }
+            let job = match self.api_client.get_job().await {
+                Ok(Some(job)) => {
+                    backoff = POLLING_INTERVAL_MS;
+                    job
+                }
+                Ok(None) => {
+                    tracing::info!("There are currently no pending batches to be proven; backing off for {} ms", backoff);
+                    tokio::time::timeout(Duration::from_millis(backoff), stop_receiver.0.changed())
+                        .await
+                        .ok();
+                    backoff = (backoff * BACKOFF_MULTIPLIER).min(MAX_BACKOFF_MS);
+                    continue;
+                }
+                Err(e) => return Err(e),
+            };
+            let (signature, batch_number, root_hash) = self.verify(*job)?;
+            let secp = Secp256k1::new();
+            let pubkey = self.signing_key.public_key(&secp);
```

Comment:
Does it make sense to make this a field to avoid recreating the public key each time?